### PR TITLE
Add comprehensive testing infrastructure setup

### PR DESCRIPTION
**Pull request by alitokgonul**
## 📌 Add Testing Infrastructure

This PR establishes a robust testing framework for the **Toad Reader Server** project, significantly improving code quality, reliability, and maintainability.

---

#### **Testing Configuration & Setup**

**`jest.config.ts`**
- Added comprehensive Jest configuration with TypeScript support
- Configured `ts-jest` preset for TypeScript transformation
- Set up test matching patterns for `__tests__/**/*.test.ts`
- Added coverage collection from `src/**/*.{ts,js}`

**`tsconfig.json` (test config)**
- Target: ES2020 with CommonJS modules
- Includes `jest` and `node` types
- Focused on `__tests__/**/*.ts` files only
- Excludes `src/` to avoid production conflicts

---

#### **Test Suites**

**`__tests__/routes/auth_routes.test.ts`** – Comprehensive authentication route testing
- 🔐 **Authentication & Authorization Tests**: login, logout, access control
- 🍪 **Cookie Management**: setcookie endpoint validation
- 📧 **Email Authentication**: passwordless flows
- 🔑 **Access Code Management**: creation, validation, usage
- ⚙️ **Admin Functions**: access code creation permissions
- 📱 **Multi-platform Support**: mobile app & web confirmations
- 🛡 **Security Features**: device login limits, session management
- 🗂 **Metadata Endpoints**: Shibboleth SSO metadata
- 🐞 **Error Handling**: extensive failure scenario testing

**`__tests__/routes/routes.test.ts`**
- Base route functionality tests
- EPUB content serving validation
- Foundation for additional route testing

---

#### **Documentation Updates**

**`README.md`**
- Clear instructions for running tests (`npm test`)
- Example test file execution

---

#### **Code Quality Improvements**
**`src/routes/auth_routes.js`**
- Added TODO comments for missing `return` statements in error handling
- Highlighted potential security issues in email validation routes

---

### 🛠 Technical Details

**Test Architecture**
- **Mocking Strategy**: Comprehensive mocking of utilities, DB connections, and external services
- **Type Safety**: Full TypeScript support with type declarations
- **Isolation**: Proper setup/teardown per test
- **Mock Factories**: Reusable for test users and IDPs
- **Error Scenarios**: Comprehensive edge-case coverage

**Coverage Areas**
- ✅ Route endpoint testing  
- ✅ Authentication flows  
- ✅ Error handling  
- ✅ Security validations  
- ✅ Database interactions (mocked)  
- ✅ Email functionality (mocked)  
- ✅ Session management  

---

### 🚀 Benefits
1. **Code Quality**: Reliable, comprehensive test coverage  
2. **Developer Confidence**: Safe refactoring with strong safety net  
3. **Documentation**: Tests as living documentation  
4. **CI/CD Ready**: Optimized Jest configuration  
5. **Type Safety**: Compile-time error detection  
6. **Maintainability**: Structured, easy-to-extend tests  

---

### 🏃 Running Tests
```bash
# Run all tests
npm test

# Run specific test file
npm test -- auth_routes

# Watch mode for development
npm test -- --watch

### CODE REVIEW

File: __tests__/routes/auth_routes.test.ts [hunk 1]
```diff
@@ -0,0 +1,1357 @@
+import express, { Express, Request, Response, NextFunction } from 'express';
+import passport from 'passport';
+import request from 'supertest';
+
+// ===== TYPE DECLARATIONS =====
+declare global {
+  // eslint-disable-next-line no-var
+  var connection: {
```
Comment (ukch) [inline __tests__/routes/auth_routes.test.ts]: Question: why `var` here?

File: __tests__/routes/routes.test.ts
Comment (ukch) [inline __tests__/routes/routes.test.ts:1]: We already discussed not renaming these files. Please can you undo this change and keep the test files alongside the files they are testing?

**Review by ukch — CHANGES_REQUESTED** [review]

File: __tests__/routes/auth_routes.test.ts (hunk 1 above)
Comment (alitokgonul) [reply __tests__/routes/auth_routes.test.ts]: Good catch! I've updated it to use the more modern TypeScript approach with interface Global instead of var.

File: __tests__/routes/routes.test.ts
Comment (alitokgonul) [reply __tests__/routes/routes.test.ts:1]: this is done

**Comment by Fydon** [comment]
We use Shibboleth rather than [email authentication](https://github.com/biblemesh/toad-reader-server/blob/be6d7b1b6ef1b5881d5a439317cc1b2308d21114/src/routes/auth_routes.js#L375-L476). As there are already tests perhaps for now no further testing of email authentication should be added, but if it starts to fail rather consider the tests and code for removal.

We also use the access code authentication to verify issues customers raise through support.

File: README.md [hunk 2]
```diff
@@ -24,11 +24,45 @@
...
+- Make sure your `.env` files are correctly populated.
+- Watch out for default setting assumptions (e.g., default language, video quality).
+- All global states should be reset between tests to prevent cross-contamination.
 
 # Updating Staging
 
-`npm run push-to-aws`
+```bash
+npm run push-to-aws
+```
```
Comment (Fydon) [inline README.md]: `push-to-aws` is a script that has been removed. It would be better to either not update this or remove the section.

File: src/routes/auth_routes.test.ts [hunk 3]
```diff
@@ -0,0 +1,1540 @@
...
+
+    it('should set cookie with correct security options', async () => {
+      const response = await request(app)
+        .get('/setcookie?cookie=test=value')
+        .expect(200);
+
+      const cookies = response.headers['set-cookie'];
+      expect(cookies).toBeDefined();
+      expect(cookies[0]).toMatch(/Max-Age=\d+/);
+      expect(cookies[0]).toMatch(/SameSite=None/i);
```
Comment (Fydon) [inline src/routes/auth_routes.test.ts:348]: Is there a reason for not checking if secure is set? In practice there may be issue if the source isn't HTTPS, but I imagine that this wouldn't impact the test of the function

File: src/routes/auth_routes.test.ts [hunk 4]
```diff
@@ -0,0 +1,1540 @@
...
+
+    it('should escape HTML in URLs for security', async () => {
+      mockUtilFunctions.escapeHTML.mockReturnValue(
+        '&lt;script&gt;alert(1)&lt;/script&gt;',
+      );
+
+      await request(app).get('/urls/example.com').expect(200);
+
+      expect(mockUtilFunctions.escapeHTML).toHaveBeenCalled();
+    });
```
Comment (Fydon) [inline src/routes/auth_routes.test.ts]: escapeHTML is called to display the URL correctly in HTML. Given that the link is passed mostly as it, if someone tried to pass a script in a URL it would be passed as is. In order for someone to cause a script to be included in the domain, they would have to able to inject calls into the person's browser, which mean the browser is already completely compromised so they could run the script directly without need to use the route this way.

In practice domains shouldn't contain characters that need HTML encoding, so if we wanted to do a security change it would be to extract the domain from the domain parameter and remove any characters requiring HTML encoding. However that is beyond the scope of this pull request.

File: src/routes/auth_routes.test.ts [hunk 5]
```diff
@@ -0,0 +1,1540 @@
...
+            callback(null, [createMockIDP()]);
+          }
+        },
+      ),
+    };
+
+    // Mock process.env
+    process.env.LOGIN_TEST_EMAIL = 'test@example.com';
+    process.env.LOGIN_TEST_CODE = 'TEST123';
+    process.env.IS_DEV = 'true';
```
Comment (Fydon) [inline src/routes/auth_routes.test.ts]: Is there a reason for setting IS_DEV? I would expect the tests to ensure that production will function correctly

File: src/routes/auth_routes.test.ts [hunk 6]
```diff
@@ -0,0 +1,1540 @@
...
+      (match) =>
+        ({
+          '&': '&amp;',
+          '<': '&lt;',
+          '>': '&gt;',
+          '"': '&quot;',
+          "'": '&#39;',
+        })[match] || match,
+    ),
+  );
```
Comment (Fydon) [inline src/routes/auth_routes.test.ts]: Why is escapeHTML being mocked to do more than util.escapeHTML [actually does](https://github.com/biblemesh/toad-reader-server/blob/50baf01f92e74c56e53c4262771a71aae7be979d/src/utils/util.js#L466-L471)?

**Review by Fydon — CHANGES_REQUESTED** [review]

File: jest.config.ts
Comment (ukch) [inline jest.config.ts:1]: Please explain why this change is necessary? I set up this file and it was working (and also type-checked)

**Comment by alitokgonul** [comment]
We'll continue this after migration is completed.

File: README.md (hunk 2 above)
Comment (alitokgonul) [reply README.md]: I have removed this

File: src/routes/auth_routes.test.ts (hunk 3 above)
Comment (alitokgonul) [reply src/routes/auth_routes.test.ts:348]: In production the cookie uses `secure: 'auto'`, [here](https://github.com/biblemesh/toad-reader-server/blob/bm-develop/src/routes/auth_routes.js#L51) so the Secure flag depends on whether the request is HTTPS. Since the unit tests run over HTTP, asserting Secure would make the test environment-dependent, so it’s intentionally not checked here.

File: src/routes/auth_routes.test.ts (hunk 4 above)
Comment (alitokgonul) [reply src/routes/auth_routes.test.ts]: Agreed, this test only checks the current rendering behavior and is not meant to provide full XSS protection.

File: src/routes/auth_routes.test.ts (hunk 5 above)
Comment (alitokgonul) [reply src/routes/auth_routes.test.ts]: yes you are right. I have removed `process.env.IS_DEV` in the test file..

File: src/routes/auth_routes.test.ts (hunk 6 above)
Comment (alitokgonul) [reply src/routes/auth_routes.test.ts]: updated the mock to reflect the current implementation.

File: jest.config.ts
Comment (alitokgonul) [reply jest.config.ts:1]: This change wasn’t required for the initial single test, but became necessary as the test suite expanded significantly.
- ensure consistent ts-jest behavior
- control test discovery scope
- isolate mocks reliably between tests
- keep coverage and imports predictable as the suite grows

The goal here is to make the test environment explicit and stable as the test surface grows.

Comment (ukch) [reply jest.config.ts:1]: OK, I'm fine with the config changes but please make the following additional changes to ensure it continues to be type-checked:

```typescript
import type { Config } from 'jest';

const config: Config = {
  // all the config comes here
};

export default config;
```
Then please check that there are no type errors before continuing.

File: src/routes/auth_routes.test.ts [hunk 7]
```diff
@@ -0,0 +1,1527 @@
+import express, { Express, Request, Response, NextFunction } from 'express';
+import passport from 'passport';
+import request from 'supertest';
+
+// ===== TYPE DECLARATIONS =====
```
Comment (ukch) [inline src/routes/auth_routes.test.ts:5]: There is an awful lot of setup in this file. A couple of questions:

1. Is all of this setup necessary? Please check (I assume it is AI-generated, which means it might be superfluous)
2. How much of this setup will be used for other modules (other route files for example)? That should be moved to a common location.

File: src/routes/auth_routes.test.ts [hunk 8]
```diff
@@ -0,0 +1,1527 @@
...
+      mockEnsureAuthenticated,
+      mockLogIn,
+      mockLog,
+    );
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+    // Clean up any lingering references
+    app = null as unknown as Express;
```
Comment (ukch) [inline src/routes/auth_routes.test.ts]: If we type app like `app: Express | undefined` we don't need to do this hackery. Then we can just do:
```suggestion
    app = undefined;
```

File: tsconfig.json
Comment (ukch) [inline tsconfig.json:1]: Is this file necessary? I was successfully using typescript before without this file, and I'm concerned that its addition here might make it more difficult to add TypeScript to the main project codebase later.

**Review by ukch — CHANGES_REQUESTED** [review]

File: src/routes/auth_routes.test.ts (hunk 7 above)
Comment (alitokgonul) [reply src/routes/auth_routes.test.ts:5]: I don't understand this comment? which file you mention about? auth_routes.test.ts ?

File: tsconfig.json
Comment (alitokgonul) [reply tsconfig.json:1]: I’ve minimized this file to avoid confusion. I think the other options I added in the tsconfig would be good to have, especially coverage, but they’re not required for the tests to run right now. For the moment, I’ve kept it as simple as possible.

File: jest.config.ts
Comment (alitokgonul) [reply jest.config.ts:1]: I don't understand this comment. where do you want me to add this code?

File: src/routes/auth_routes.test.ts (hunk 7 above)
Comment (ukch) [reply src/routes/auth_routes.test.ts:5]: Sorry, which part don't you understand. I'm talking about `auth_routes.test.ts` - look at it and there is a ton of setup code. Is this AI generated? Please can you do a review of the setup code in this module and double check what is actually required, and also which parts should be reused in other test files later.

File: jest.config.ts
Comment (ukch) [reply jest.config.ts:1]: Please look at the diff for this file and it will be obvious what I'm talking about.

Comment (ukch) [reply jest.config.ts:1]: Actually please just undo all changes you have made to this file. I don't understand why you just deleted all the extra config (that you were previously saying was necessary) but please just put it back to how it is on `bm-develop`.

File: tsconfig.json
Comment (ukch) [reply tsconfig.json:1]: Please do the following:

* Delete this file
* Run tests
* Check status:
  * If tests fail, restore the file and run tests again. The file is necessary.
  * If tests pass, keep the file deleted. It is not necessary.

**Review by ukch — CHANGES_REQUESTED** [review]
I don't really understand your latest changes and how they relate to my PR review. I am suggesting you revert some of the config changes (see comments) and I think the tests will likely still work fine without them.

File: tsconfig.json
Comment (alitokgonul) [reply tsconfig.json:1]: the file is neccessary so I restored the file

File: src/routes/auth_routes.test.ts (hunk 7 above)
Comment (alitokgonul) [reply src/routes/auth_routes.test.ts:5]: I simplified it, and removed unnecessary parts

File: jest.config.ts [hunk 9]
```diff
@@ -12,4 +12,4 @@ const config: Config = {
   },
 };
 
-export default config;
+export default config;
```
Comment (ukch) [inline jest.config.ts:15]: Please double-check that your pre-commit linting is running. This file is badly-formatted where it was not before.

**Review by ukch — APPROVED** [review]
One minor comment, otherwise this is good to merge

**Comment by ukch** [comment]
Also please merge `bm-develop` and ensure lint/test/build fully passes, since this is now fixed and Node version is also upgraded.

File: src/routes/auth_routes.test.ts (hunk 4 above)
Comment (Fydon) [reply src/routes/auth_routes.test.ts]: As this test is checking if escapeHTML is called, please rename the test to something like `should call escapeHTML` and change `mockReturnValue` to something simple like `foo`. Otherwise if it should test the output value, the actual escapeHTML should be called and the output be tested.

**Review by Fydon — CHANGES_REQUESTED** [review]

File: src/routes/auth_routes.test.ts (hunk 6 above)
Comment (Fydon) [reply src/routes/auth_routes.test.ts]: I'll leave this as unresolved until the other comment is resolved, given that they are related. From what I see the updated code has `mockUtilFunctions.escapeHTML.mockImplementation((text: string) => text);` which is not the same as what `escapeHTML` does, but if nothing is checking the output then it doesn't matter.

**Review by Fydon — APPROVED** [review]
Thank you
